doc/examples/boilerplate.py

```diff
@@ -0,0 +1,135 @@
+"""
+A docstring
+
+This can double as a helpful message which explains how the program is run.
+"""
+
+from __future__ import absolute_import, division, print_function
+
+import logging
+import sys
+from typing import List  # noqa: F401  Flake8 doesn't recognise Python 2 typing.
+
+import libtbx.phil
+from dials.array_family import flex
+import dials.util
+import dials.util.log
+from dials.util.options import OptionParser, flatten_reflections, flatten_experiments
+from dxtbx.model import ExperimentList
+
+
+# Define a logger
+logger = logging.getLogger("dials.logger_name")
+
+# Define the master PHIL scope for this program
+phil_scope = libtbx.phil.parse(
+    """
+    output {
+        reflections = stronger.refl
+            .type = path
+            .help = "Help strings are helpful, wherever possible."
+        log = dials.command_name.log
+            .type = path
+    }
+    bool_parameter = False
+      .type = bool
+    integer_parameter = 0
+      .type = int
+    """
+)
+
+
+def do_stuff(experiments, reflections, params):
+    # type: (ExperimentList, flex.reflection_table, libtbx.phil.scope_extract) -> None
+    """
+    Write the behaviour of the program as functions and classes outside run().
+
+    Don't include file output here, remember that this function may be re-used
+    elsewhere by someone who doesn't need the output written immediately to file.
+
+    It can be especially helpful to document any expected exceptions that might be
+    raised, in order to keep track of what needs to be handled in any code that
+    re-uses this function.
+
+    Args:
+        experiments:  An experiment list.
+        reflections:  A reflection table.
+        params:       Some parameters, in the form of a scope_extract object,
+                      which is the usable form of a parsed PHIL scope.
+
+    Raises:
+        RuntimeError:  if you pass an empty experiment list.
+    """
+    logger.info("Hello world!")
+
+    # Here's an example of an error that might be raised, as documented above.
+    if not experiments:
+        raise RuntimeError
+
+    logger.info("The input reflection table contains %d reflections.", len(reflections))
+    logger.info(
+        "The input experiment list contains %d imagesets.", len(experiments.imagesets)
+    )
+
+    logger.info("integer_parameter: %i", params.integer_parameter)
+    logger.info("bool_parameter: %s", params.bool_parameter)
+
+
+def run(args=None, phil=phil_scope):  # type: (List[str], libtbx.phil.scope) -> None
```

Comment:
default to `args=` not set to `sys.argv[1:]`

Reply:
If `OptionParser` gets passed None it defaults to sys.argv. Otherwise, you run the risk of list-caching, the same as putting a list object as a default to a function parameter (e.g. errors with `def a(some=[]): some.append(x)`)

Reply:
Yep, the default behaviour is obfuscated by the DIALS option parser inheriting defaults from the (old) standard optparse option parser, but not documenting that clearly so nobody realises.  I'm happy to talk about ripping up the DIALS option parser, if you want...  :-P